File: S3TransferManager-Sample/Objective-C/S3TransferManagerSample/DownloadViewController.m
/*
 * Copyright 2010-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

#import "DownloadViewController.h"

#import "S3.h"
#import "JTSImageViewController.h"
#import "Constants.h"

@interface DownloadViewController ()

@property (nonatomic, strong) NSMutableArray *collection;

@end

@implementation DownloadViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.collection = [NSMutableArray new];
    [self listObjects:self];

    NSError *error = nil;
    if (![[NSFileManager defaultManager] createDirectoryAtPath:[NSTemporaryDirectory() stringByAppendingPathComponent:@"download"]
                                   withIntermediateDirectories:YES
                                                    attributes:nil
                                                         error:&error]) {
        NSLog(@"Creating 'download' directory failed. Error: [%@]", error);
    }
}

#pragma mark - User action methods

- (IBAction)showAlertController:(id)sender {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Available Actions"
                                                                             message:@"Choose your action."
                                                                      preferredStyle:UIAlertControllerStyleActionSheet];

    __weak DownloadViewController *weakSelf = self;
    UIAlertAction *refreshAction = [UIAlertAction actionWithTitle:@"Refresh"
                                                            style:UIAlertActionStyleDefault
                                                          handler:^(UIAlertAction *action) {
                                                              DownloadViewController *strongSelf = weakSelf;
                                                              [strongSelf.collection removeAllObjects];
                                                              [strongSelf listObjects:self];
                                                          }];
    [alertController addAction:refreshAction];

    UIAlertAction *downloadAllAction = [UIAlertAction actionWithTitle:@"Download All"
                                                                style:UIAlertActionStyleDefault
                                                              handler:^(UIAlertAction *action) {
                                                                  DownloadViewController *strongSelf = weakSelf;
                                                                  [strongSelf downloadAll];
                                                              }];
    [alertController addAction:downloadAllAction];

    UIAlertAction *cancelAllDownloadsAction = [UIAlertAction actionWithTitle:@"Cancel All Downloads"
                                                                       style:UIAlertActionStyleDefault
                                                                     handler:^(UIAlertAction *action) {
                                                                         DownloadViewController *strongSelf = weakSelf;
                                                                         [strongSelf cancelAllDownloads:self];
                                                                     }];
    [alertController addAction:cancelAllDownloadsAction];

    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel"
                                                           style:UIAlertActionStyleCancel
                                                         handler:nil];
    [alertController addAction:cancelAction];

    [self presentViewController:alertController
                       animated:YES
                     completion:nil];
}

- (void)listObjects:(id)sender {
    AWSS3 *s3 = [AWSS3 defaultS3];

    AWSS3ListObjectsRequest *listObjectsRequest = [AWSS3ListObjectsRequest new];
    listObjectsRequest.bucket = S3BucketName;
    [[s3 listObjects:listObjectsRequest] continueWithBlock:^id(BFTask *task) {
        if (task.error) {
            NSLog(@"listObjects failed: [%@]", task.error);
        } else {
            AWSS3ListObjectsOutput *listObjectsOutput = task.result;
            for (AWSS3Object *s3Object in listObjectsOutput.contents) {
                NSString *downloadingFilePath = [[NSTemporaryDirectory() stringByAppendingPathComponent:@"download"] stringByAppendingPathComponent:s3Object.key];
                NSURL *downloadingFileURL = [NSURL fileURLWithPath:downloadingFilePath];

                if ([[NSFileManager defaultManager] fileExistsAtPath:downloadingFilePath]) {
                    [self.collection addObject:downloadingFileURL];
                } else {
                    AWSS3TransferManagerDownloadRequest *downloadRequest = [AWSS3TransferManagerDownloadRequest new];
                    downloadRequest.bucket = S3BucketName;
                    downloadRequest.key = s3Object.key;
                    downloadRequest.downloadingFileURL = downloadingFileURL;
                    [self.collection addObject:downloadRequest];
                }
            }

            dispatch_async(dispatch_get_main_queue(), ^{
                [self.collectionView reloadData];
            });
        }
        return nil;
    }];
}

- (void)download:(AWSS3TransferManagerDownloadRequest *)downloadRequest {
    switch (downloadRequest.state) {
        case AWSS3TransferManagerRequestStateNotStarted:
        case AWSS3TransferManagerRequestStatePaused:
        {
            AWSS3TransferManager *transferManager = [AWSS3TransferManager defaultS3TransferManager];
            [[transferManager download:downloadRequest] continueWithBlock:^id(BFTask *task) {
                if ([task.error.domain isEqualToString:AWSS3TransferManagerErrorDomain]
                    && task.error.code == AWSS3TransferManagerErrorPaused) {
                    NSLog(@"Download paused.");
                } else if (task.error) {
                    NSLog(@"Upload failed: [%@]", task.error);
                } else {
                    __weak DownloadViewController *weakSelf = self;
                    dispatch_async(dispatch_get_main_queue(), ^{
                        DownloadViewController *strongSelf = weakSelf;

                        NSUInteger index = [strongSelf.collection indexOfObject:downloadRequest];
                        [strongSelf.collection replaceObjectAtIndex:index
                                                         withObject:downloadRequest.downloadingFileURL];
                        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:index
                                                                    inSection:0];
                        [strongSelf.collectionView reloadItemsAtIndexPaths:@[indexPath]];
                    });
                }
                return nil;
            }];
        }
            break;
        default:
            break;
    }
}

- (void)downloadAll {
    [self.collection enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        if ([obj isKindOfClass:[AWSS3TransferManagerDownloadRequest class]]) {
            AWSS3TransferManagerDownloadRequest *downloadRequest = obj;
            if (downloadRequest.state == AWSS3TransferManagerRequestStateNotStarted
                || downloadRequest.state == AWSS3TransferManagerRequestStatePaused) {
                [self download:downloadRequest];
            }
        }
    }];

    [self.collectionView reloadData];
}

- (void)cancelAllDownloads:(id)sender {
    [self.collection enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        if ([obj isKindOfClass:[AWSS3TransferManagerDownloadRequest class]]) {
            AWSS3TransferManagerDownloadRequest *downloadRequest = obj;
            if (downloadRequest.state == AWSS3TransferManagerRequestStateRunning
                || downloadRequest.state == AWSS3TransferManagerRequestStatePaused) {
                [[downloadRequest cancel] continueWithBlock:^id(BFTask *task) {
                    if (task.error) {
                        NSLog(@"The cancel request failed: [%@]", task.error);
                    }
                    return nil;
                }];
            }
        }
    }];

    [self.collectionView reloadData];
}

#pragma mark - Collection View methods

- (NSInteger)collectionView:(UICollectionView *)collectionView
     numberOfItemsInSection:(NSInteger)section {
    return [self.collection count];
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    id object = [self.collection objectAtIndex:indexPath.row];
    DownloadCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"DownloadCollectionViewCell"
                                                                                 forIndexPath:indexPath];

    if ([object isKindOfClass:[AWSS3TransferManagerDownloadRequest class]]) {
        AWSS3TransferManagerDownloadRequest *downloadRequest = object;
        downloadRequest.downloadProgress = ^(int64_t bytesWritten, int64_t totalBytesWritten, int64_t totalBytesExpectedToWrite) {
            dispatch_async(dispatch_get_main_queue(), ^{
                if (totalBytesExpectedToWrite > 0) {
                    cell.progressView.progress = (float)((double) totalBytesWritten / totalBytesExpectedToWrite);
                }
            });
        };

        cell.label.hidden = NO;
        cell.imageView.image = nil;
        switch (downloadRequest.state) {
            case AWSS3TransferManagerRequestStateNotStarted:
            case AWSS3TransferManagerRequestStatePaused:
                cell.progressView.progress = 0.0f;
                cell.label.text = @"Download";
                break;

            case AWSS3TransferManagerRequestStateRunning:
                cell.label.text = @"Pause";
                break;

            case AWSS3TransferManagerRequestStateCanceling:
                cell.progressView.progress = 1.0f;
                cell.label.text = @"Cancelled";
                break;

            default:
                break;
        }
    } else if ([object isKindOfClass:[NSURL class]]) {
        cell.label.hidden = YES;
        NSURL *downloadFileURL = object;
        cell.progressView.progress = 1.0f;
        cell.imageView.image = [UIImage imageWithData:[NSData dataWithContentsOfURL:downloadFileURL]];
    }

    return cell;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    [collectionView deselectItemAtIndexPath:indexPath animated:YES];
    id object = [self.collection objectAtIndex:indexPath.row];

    if ([object isKindOfClass:[AWSS3TransferManagerDownloadRequest class]]) {
        AWSS3TransferManagerDownloadRequest *downloadRequest = object;

        switch (downloadRequest.state) {
            case AWSS3TransferManagerRequestStateNotStarted:
            case AWSS3TransferManagerRequestStatePaused:
                [self download:downloadRequest];
                break;

            case AWSS3TransferManagerRequestStateRunning:
            {
                [[downloadRequest pause] continueWithBlock:^id(BFTask *task) {
                    if (task.error) {
                        NSLog(@"The pause request failed: [%@]", task.error);
                    } else {
                        __weak DownloadViewController *weakSelf = self;
                        dispatch_async(dispatch_get_main_queue(), ^{
                            DownloadViewController *strongSelf = weakSelf;

                            NSUInteger index = [strongSelf.collection indexOfObject:downloadRequest];
                            NSIndexPath *indexPath = [NSIndexPath indexPathForRow:index
                                                                        inSection:0];
                            [strongSelf.collectionView reloadItemsAtIndexPaths:@[indexPath]];
                        });
                    }
                    return nil;
                }];
            }
                break;

            default:
                break;
        }

        [self.collectionView reloadData];
    } else if ([object isKindOfClass:[NSURL class]]) {
        NSURL * downloadingFileURL = object;

        JTSImageInfo *imageInfo = [JTSImageInfo new];
        imageInfo.image = [UIImage imageWithData:[NSData dataWithContentsOfURL:downloadingFileURL]];

        JTSImageViewController *imageViewer = [[JTSImageViewController alloc]
                                               initWithImageInfo:imageInfo
                                               mode:JTSImageViewControllerMode_Image
                                               backgroundStyle:JTSImageViewControllerBackgroundOption_Blurred];
        [imageViewer showFromViewController:self
                                 transition:JTSImageViewControllerTransition_FromOffscreen];
    }
}

@end

@implementation DownloadCollectionViewCell

@end

File: S3TransferManager-Sample/Swift/S3TransferManagerSampleSwift/DownloadViewController.swift
/*
* Copyright 2010-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
*
* Licensed under the Apache License, Version 2.0 (the "License").
* You may not use this file except in compliance with the License.
* A copy of the License is located at
*
*  http://aws.amazon.com/apache2.0
*
* or in the "license" file accompanying this file. This file is distributed
* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
* express or implied. See the License for the specific language governing
* permissions and limitations under the License.
*/

import UIKit

class DownloadViewController: UIViewController, UICollectionViewDelegate, UICollectionViewDataSource {
    
    @IBOutlet weak var collectionView: UICollectionView!
    
    var downloadRequests = Array<AWSS3TransferManagerDownloadRequest?>()
    var downloadFileURLs = Array<NSURL?>()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        listObjects()
        let error = NSErrorPointer()
        do {
            try NSFileManager.defaultManager().createDirectoryAtURL(
                NSURL(fileURLWithPath: NSTemporaryDirectory()).URLByAppendingPathComponent("download"),
                withIntermediateDirectories: true,
                attributes: nil)
        } catch let error1 as NSError {
            error.memory = error1
            print("Creating 'download' directory failed. Error: \(error)")
        }
    }
    
    @IBAction func showAlertController(barButtonItem: UIBarButtonItem) {
        let alertController = UIAlertController(
            title: "Available Actions",
            message: "Choose your action.",
            preferredStyle: .ActionSheet)
        
        let refreshAction = UIAlertAction(
            title: "Refresh",
            style: .Default) { (action) -> Void in
                self.downloadRequests.removeAll(keepCapacity: false)
                self.downloadFileURLs.removeAll(keepCapacity: false)
                self.collectionView.reloadData()
                self.listObjects()
        }
        alertController.addAction(refreshAction)
        
        let downloadAllAction = UIAlertAction(
            title: "Download All",
            style: .Default) { (action) -> Void in
                self.downloadAll()
        }
        alertController.addAction(downloadAllAction)
        
        let cancelAllDownloadsAction = UIAlertAction(
            title: "Cancel All Downloads",
            style: .Default) { (action) -> Void in
                self.cancelAllDownloads()
        }
        alertController.addAction(cancelAllDownloadsAction)
        
        let cancelAction = UIAlertAction(
            title: "Cancel",
            style: .Cancel,
            handler: nil)
        alertController.addAction(cancelAction)
        
        self.presentViewController(
            alertController,
            animated: true,
            completion: nil)
    }
    
    func listObjects() {
        let s3 = AWSS3.defaultS3()
        
        let listObjectsRequest = AWSS3ListObjectsRequest()
        listObjectsRequest.bucket = S3BucketName
        s3.listObjects(listObjectsRequest).continueWithBlock { (task) -> AnyObject! in
            if let error = task.error {
                print("listObjects failed: [\(error)]")
            }
            if let exception = task.exception {
                print("listObjects failed: [\(exception)]")
            }
            if let listObjectsOutput = task.result as? AWSS3ListObjectsOutput {
                if let contents = listObjectsOutput.contents as? [AWSS3Object] {
                    for s3Object in contents {
                        let downloadingFileURL = NSURL(fileURLWithPath: NSTemporaryDirectory()).URLByAppendingPathComponent("download").URLByAppendingPathComponent(s3Object.key)
                        let downloadingFilePath = downloadingFileURL.path!
                        
                        if NSFileManager.defaultManager().fileExistsAtPath(downloadingFilePath) {
                            self.downloadRequests.append(nil)
                            self.downloadFileURLs.append(downloadingFileURL)
                        } else {
                            let downloadRequest = AWSS3TransferManagerDownloadRequest()
                            downloadRequest.bucket = S3BucketName
                            downloadRequest.key = s3Object.key
                            downloadRequest.downloadingFileURL = downloadingFileURL
                            
                            self.downloadRequests.append(downloadRequest)
                            self.downloadFileURLs.append(nil)
                        }
                    }
                    
                    dispatch_async(dispatch_get_main_queue(), { () -> Void in
                        self.collectionView.reloadData()
                    })
                }
            }
            return nil
        }
    }
    
    func download(downloadRequest: AWSS3TransferManagerDownloadRequest) {
        switch (downloadRequest.state) {
        case .NotStarted, .Paused:
            let transferManager = AWSS3TransferManager.defaultS3TransferManager()
            transferManager.download(downloadRequest).continueWithBlock({ (task) -> AnyObject! in
                if let error = task.error {
                    if error.domain == AWSS3TransferManagerErrorDomain as String
                        && AWSS3TransferManagerErrorType(rawValue: error.code) == AWSS3TransferManagerErrorType.Paused {
                            print("Download paused.")
                    } else {
                        print("download failed: [\(error)]")
                    }
                } else if let exception = task.exception {
                    print("download failed: [\(exception)]")
                } else {
                    dispatch_async(dispatch_get_main_queue(), { () -> Void in
                        if let index = self.indexOfDownloadRequest(self.downloadRequests, downloadRequest: downloadRequest) {
                            self.downloadRequests[index] = nil
                            self.downloadFileURLs[index] = downloadRequest.downloadingFileURL
                            
                            let indexPath = NSIndexPath(forRow: index, inSection: 0)
                            self.collectionView.reloadItemsAtIndexPaths([indexPath])
                        }
                    })
                }
                return nil
            })
            
            break
        default:
            break
        }
    }
    
    func downloadAll() {
        for (_, value) in self.downloadRequests.enumerate() {
            if let downloadRequest = value {
                if downloadRequest.state == .NotStarted
                    || downloadRequest.state == .Paused {
                        self.download(downloadRequest)
                }
            }
        }
        
        self.collectionView.reloadData()
    }
    
    func cancelAllDownloads() {
        for (_, value) in self.downloadRequests.enumerate() {
            if let downloadRequest = value {
                if downloadRequest.state == .Running
                    || downloadRequest.state == .Paused {
                        downloadRequest.cancel().continueWithBlock({ (task) -> AnyObject! in
                            if let error = task.error {
                                print("cancel() failed: [\(error)]")
                            } else if let exception = task.exception {
                                print("cancel() failed: [\(exception)]")
                            }
                            return nil
                        })
                }
            }
        }
        
        self.collectionView.reloadData()
    }
    
    func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.downloadRequests.count
    }
    
    func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCellWithReuseIdentifier("DownloadCollectionViewCell", forIndexPath: indexPath) as! DownloadCollectionViewCell
        
        if let downloadRequest = self.downloadRequests[indexPath.row] {
            downloadRequest.downloadProgress = { (bytesWritten, totalBytesWritten, totalBytesExpectedToWrite) -> Void in
                dispatch_async(dispatch_get_main_queue(), { () -> Void in
                    if totalBytesExpectedToWrite > 0 {
                        cell.progressView.progress = Float(Double(totalBytesWritten) / Double(totalBytesExpectedToWrite));
                    }
                })
            }
            cell.label.hidden = false
            cell.imageView.image = nil
            
            switch downloadRequest.state {
            case .NotStarted, .Paused:
                cell.progressView.progress = 0.0
                cell.label.text = "Download"
                break
                
            case .Running:
                cell.label.text = "Pause"
                break
                
            case .Canceling:
                cell.progressView.progress = 1.0
                cell.label.text = "Cancelled"
                break
                
            default:
                break
            }
        }
        
        if let downloadFileURL = self.downloadFileURLs[indexPath.row] {
            cell.label.hidden = true
            cell.progressView.progress = 1.0
            if let data = NSData(contentsOfURL: downloadFileURL) {
                cell.imageView.image = UIImage(data: data)
            }
        }
        
        return cell
    }
    
    func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
        collectionView.deselectItemAtIndexPath(indexPath, animated: true)
        if let downloadRequest = self.downloadRequests[indexPath.row] {
            
            switch (downloadRequest.state) {
            case .NotStarted, .Paused:
                self.download(downloadRequest)
                break
                
            case .Running:
                downloadRequest.pause().continueWithBlock({ (task) -> AnyObject! in
                    if let error = task.error {
                        print("pause() failed: [\(error)]")
                    } else if let exception = task.exception {
                        print("pause() failed: [\(exception)]")
                    } else {
                        dispatch_async(dispatch_get_main_queue(), { () -> Void in
                            collectionView.reloadItemsAtIndexPaths([indexPath])
                        })
                    }
                    return nil
                })
                break
                
            default:
                break
            }
            
            collectionView.reloadData()
        }
        
        if let downloadFileURL = self.downloadFileURLs[indexPath.row] {
            if let data = NSData(contentsOfURL: downloadFileURL) {
                let imageInfo = JTSImageInfo()
                imageInfo.image = UIImage(data: data)
                
                let imageViewer = JTSImageViewController(
                    imageInfo: imageInfo,
                    mode: .Image,
                    backgroundStyle: .Blurred)
                imageViewer.showFromViewController(self, transition: .FromOffscreen)
            }
        }
    }
    
    func indexOfDownloadRequest(array: Array<AWSS3TransferManagerDownloadRequest?>, downloadRequest: AWSS3TransferManagerDownloadRequest?) -> Int? {
        for (index, object) in array.enumerate() {
            if object == downloadRequest {
                return index
            }
        }
        return nil
    }
}

class DownloadCollectionViewCell: UICollectionViewCell {
    @IBOutlet weak var progressView: UIProgressView!
    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var label: UILabel!
}
